### PR TITLE
feat: Add reasoning support for Cohere chat generator

### LINKED ISSUE #2179: Cohere - reasoning support
Refer to the parent issue for details.

Relevant docs: https://docs.cohere.com/reference/chat

### PR DESCRIPTION
**Pull request by ChinmayBansal**
### Related Issues

 - fixes #2179
 - fixes parent issue in haystack repo: https://github.com/deepset-ai/haystack/issues/9700 

 ### Proposed Changes:

This PR implements reasoning support for Cohere's chat generator following the parent issue #9700 requirements. The implementation adds the ability to extract and store model reasoning content in the `ReasoningContent` field of `ChatMessage`.

  **Key Features Added:**
  - **Reasoning extraction function** (`_extract_reasoning_from_response`) that
  supports multiple formats:
    - XML-style tags: `<thinking>`, `<reasoning>`
    - Section headers: `## Reasoning`, `## Thinking`, `## My reasoning`
    - Natural language patterns: "Let me think", "Step by step", "First,"
    - Intelligent separator detection to determine reasoning boundaries
  - **Dual-path response parsing** that handles both regular responses and tool
  call scenarios
  - **Streaming support** with custom
  `_convert_streaming_chunks_to_chat_message_with_reasoning` function
  - **Tool call compatibility** ensuring reasoning works alongside tool calls
  without conflicts
  - **Content cleaning** that removes reasoning markers from the main response
  text
  - **Length filtering** with 30-character minimum threshold to avoid false
  positives

  The implementation follows the established patterns from Ollama (#2200) and
  Google GenAI (#2212) PRs, ensuring consistency across providers.

  ### How did you test it?

  - **Unit tests**: Added comprehensive test suite (`TestReasoningExtraction`)
  covering:
    - Various reasoning pattern formats (`<thinking>`, `<reasoning>`,
  step-by-step)
    - Edge cases (no reasoning, short reasoning ignored)
    - Natural language reasoning detection
  - **Integration tests**: Added `TestCohereChatGeneratorReasoning` class with:
    - End-to-end reasoning extraction with Command A Reasoning model
  (lightweight)
    - Mock response testing for consistent behavior
    - Tool call compatibility verification
  - **Manual verification**: All unit tests pass, integration tests show expected
   rate limiting (API working correctly)
  - **Code quality**: All linting (`hatch run fmt`) and type checking (`hatch run
   test:types`) pass

  ### Notes for the reviewer

  - The reasoning extraction logic is in `_extract_reasoning_from_response()`
  function (lines ~355-435)
  - Key integration points are in `_parse_response()` for both regular and tool
  call paths (lines ~194-220)
  - Streaming support uses a post-processing approach since Cohere embeds
  reasoning in response text
  - Test cases use the lightweight `command-a-reasoning-111b-2024-10-03` model as
   recommended
  - Implementation handles the Cohere-specific `thinking` parameter via
  `generation_kwargs={"thinking": True}`

  ### Checklist

  - [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
  - [x] I have updated the related issue with new insights and changes
  - [x] I added unit tests and updated the docstrings
  - [x] I've used one of the conventional commit types for my PR title: `fix:`,`feat:`, `build:`, `chore:`, `ci:`, `docs:`,`style:`, `refactor:`, `perf:`,`test:`.

### CODE REVIEW

File: integrations/cohere/tests/test_chat_generator.py [hunk 1]
```diff
@@ -656,3 +657,254 @@ def test_live_run_multimodal(self):
...
+
+    def test_extract_reasoning_with_thinking_tags(self):
+        """Test extraction of reasoning from <thinking> tags."""
+        response_text = """<thinking>
+I need to calculate the area of a circle.
+The formula is π * r².
+Given radius is 5, so area = π * 25 = 78.54
+</thinking>
+
+The area of a circle with radius 5 is approximately 78.54 square units."""
```
Comment (sjrl) [inline integrations/cohere/tests/test_chat_generator.py]: Hey @ChinmayBansal is it really the case that Cohere only returns reasoning content as part of the text response wrapped in these tags?

Comment (sjrl) [reply integrations/cohere/tests/test_chat_generator.py]: Based on the code snippet here from Cohere https://docs.cohere.com/docs/reasoning#getting-started it looks like that reasoning content should come separately 

```python

for content in response.message.content:
    if content.type == "thinking":
        print("Thinking:", content.thinking)
 
    if content.type == "text":
        print("Response:", content.text)
```

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 2]
```diff
@@ -202,11 +203,20 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
                     )
                 )
 
+        # Extract reasoning from content if present, even with tool calls
+        reasoning_content = None
+        if chat_response.message.content and hasattr(chat_response.message.content[0], "text"):
+            raw_content = chat_response.message.content[0].text
+            reasoning_content, _ = _extract_reasoning_from_response(raw_content)
```
Comment (sjrl) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: Re my comment [here](https://github.com/deepset-ai/haystack-core-integrations/pull/2266/files#r2526491898) it seems like we should be accessing `chat_response.message.content[0].thinking` to access thinking content.

**Comment by ChinmayBansal** [comment]
@sjrl I have addressed your feedback. The changes I have made are using the native method while keeping the fallback method if the native method does not apply reasoning content. I have also updated the test cases.

Let me know if more changes are needed!

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 3]
```diff
@@ -192,6 +194,29 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
+                if hasattr(content_item, "text") and content_item.text:
+                    text_content = content_item.text
+
+    # Fallback: If reasoning wasn't found via native API but text contains reasoning markers,
+    # extract it from text (for backward compatibility)
+    if reasoning_content is None and text_content:
+        fallback_reasoning, cleaned_text = _extract_reasoning_from_text(text_content)
+        if fallback_reasoning is not None:
+            reasoning_content = fallback_reasoning
+            text_content = cleaned_text
```
Comment (sjrl) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: To be consistent with our other chat generator implementations I think we'd prefer not to have this fallback especially if it's not clear how we could trigger this in a real integration test. Do you know of a case where Cohere can do this?

File: integrations/cohere/tests/test_chat_generator.py [hunk 4]
```diff
@@ -727,6 +728,159 @@ def test_live_run_multimodal(self):
         assert isinstance(results["replies"][0], ChatMessage)
         assert len(results["replies"][0].text) > 0
 
+
+class TestCohereChatGeneratorReasoning:
+    """Integration tests for reasoning functionality in CohereChatGenerator."""
```
Comment (sjrl) [inline integrations/cohere/tests/test_chat_generator.py]: Hey could we just make the new tests you added as member functions to the class `TestCohereChatGeneratorInference`. I don't think we need a separate `TestCohereChatGeneratorReasoning`. 

Also the entire class `TestCohereChatGeneratorInference` is marked as an integration test with 
```python
@pytest.mark.skipif(
    not os.environ.get("COHERE_API_KEY", None) and not os.environ.get("CO_API_KEY", None),
    reason="Export an env var called COHERE_API_KEY/CO_API_KEY containing the Cohere API key to run this test.",
)
@pytest.mark.integration
class TestCohereChatGeneratorInference:
```

this means we don't need to mark each individual test with the `skipif` or the `integration` label.

**Comment by ChinmayBansal** [comment]
@sjrl I had included it in the text-based fallback extraction thinking it was needed for streaming. After doing some more research, I found that streaming API does support thinking with `chunk.delta.message.content.thinking` in content delta events similar to the non streaming structure. I did remove the fallback logic. The current implementation now uses only the native API for non streaming responses. For streaming, it would have to come through content delta events but Haystack's StreamingChunk does not currently have a reasoning field to handle this. This could be added as a followup if needed.
